servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/WriteSingleSubscriber.java

```diff
@@ -51,21 +58,29 @@ public void onSubscribe(Cancellable cancellable) {
 
     @Override
     public void onSuccess(@Nullable Object result) {
-        // If we are not on the EventLoop then both the write and the flush will be enqueued on the EventLoop so
-        // ordering should be correct.
-        channel.writeAndFlush(result).addListener(future -> {
-            Throwable cause = future.cause();
-            if (cause == null) {
-                notifyComplete();
-            } else {
-                notifyError(cause);
-            }
-        });
+        if (terminatedUpdater.compareAndSet(this, AWAITING_RESULT, TERMINATED)) {
+            // If we are not on the EventLoop then both the write and the flush will be enqueued on the EventLoop so
+            // ordering should be correct.
+            channel.writeAndFlush(result).addListener(future -> {
+                Throwable cause = future.cause();
+                if (cause == null) {
+                    notifyComplete();
+                } else {
+                    notifyError(cause);
+                }
+            });
+        } else {
+            LOGGER.error("Failed to write data after permature close of the WriteListener.");
```

Comment:
Also log the `result`?

```suggestion
            LOGGER.error("Ignoring write {} as the listener is already closed.", result);
```